zap/src/output/client.rs

```diff
@@ -1,260 +1,363 @@
+			file,
+			buff: String::new(),
+			tabs: 0,
+		}
+	}
+
+	fn push_tydecl(&mut self, tydecl: &TyDecl) {
+		let name = &tydecl.name;
+		let ty = &tydecl.ty;
+
+		self.push_line(&format!("export type {name} = {ty}"));
+
+		self.push_line(&format!("function types.write_{name}(value: {name})"));
+		self.push_stmts(&gen_ser(ty, "value".into(), self.file.write_checks));
+		self.push_line("end");
+
+		self.push_line(&format!("function types.read_{name}()"));
+		self.push_line("local value;");
+		self.push_stmts(&gen_des(ty, "value".into(), false));
+		self.push_line("return value");
+		self.push_line("end");
+	}
+
+	fn push_tydecls(&mut self) {
+		for tydecl in self.file.ty_decls.iter() {
+			self.push_tydecl(tydecl);
+		}
+	}
+
+	fn push_reliable_header(&mut self) {
+		self.push_line("reliable.OnClientEvent:Connect(function(buff, inst)");
+		self.indent();
+		self.push_line("incoming_buff = buff");
+		self.push_line("incoming_inst = inst");
+		self.push_line("incoming_read = 0");
 
-impl<'a> Display for ClientFile<'a> {
-	fn fmt(&self, f: &mut std::fmt::Formatter<'_>) -> std::fmt::Result {
-		let file = self.0;
+		self.push_line("local len = buffer.len(buff)");
+		self.push_line("while incoming_read < len do");
 
-		if file.ev_decls.is_empty() {
-			return Ok(());
+		self.indent();
+
+		self.push_line(&format!(
+			"local id = buffer.read{}(buff, read({}))",
+			self.file.event_id_ty(),
+			self.file.event_id_ty().size()
+		));
+	}
+	fn push_reliable_callback(&mut self, first: bool, ev: &EvDecl, id: usize) {
+		self.push_indent();
+
+		if first {
+			self.push("if ");
+		} else {
+			self.push("elseif ");
 		}
 
-		writeln!(f, "--!native")?;
-		writeln!(f, "--!optimize 2")?;
-		writeln!(f, "-- Generated by zap v{}", env!("CARGO_PKG_VERSION"))?;
+		self.push(&format!("id == {id} then"));
+		self.push("\n");
 
-		write!(f, "{}", include_str!("base.luau"))?;
-		write!(f, "{}", include_str!("client.luau"))?;
+		self.indent();
 
-		writeln!(f, "local events = table.create({})", file.ev_decls.len())?;
+		self.push_line("local value");
+		self.push_stmts(&gen_des(&ev.data, "value".into(), false));
 
-		// Output Types
-		for ty_decl in file.ty_decls.iter() {
-			write!(f, "{}", ty_decl)?;
+		match ev.call {
+			EvCall::SingleSync => self.push_line(&format!("if events[{id}] then events[{id}](value) end")),
+			EvCall::SingleAsync => self.push_line(&format!("if events[{id}] then task.spawn(events[{id}], value) end")),
+			EvCall::ManySync => self.push_line(&format!("for _, cb in events[{id}] do cb(value) end")),
+			EvCall::ManyAsync => self.push_line(&format!("for _, cb in events[{id}] do task.spawn(cb, value) end")),
 		}
 
-		// Output Reliable Callbacks
-		if file
+		self.dedent();
+	}
+
+	fn push_reliable_footer(&mut self) {
+		self.push_line("else");
+		self.indent();
+		self.push_line("error(\"Unknown event id\")");
+		self.dedent();
+		self.push_line("end");
+		self.dedent();
+		self.push_line("end");
+		self.dedent();
+		self.push_line("end)");
```

Comment:
Maybe this should use `self.tabs and from that decide how many "ends" it should insert?

Reply:
It's always going to be 3 because of the above code - general rule is that when you explicitly open a scope you must explicitly close it somewhere else.

Reply:
Sounds good - I agree.